### PR TITLE
normalize ranged len read

### PR DESCRIPTION
**Pull request by noahcape**
Given a label, and the range which it should satisfy, this will normalize the label while not allowing collisions. 

Using `log2_roundup` of ranged distance and the amount of needed padding to satisfy normalization.

Using nucleotide to Integer mapping
A: 0
C: 1
T: 2
G: 3

And interpreting them as bits relating to the right most 2 bits of the length to normalize of the label.

### CODE REVIEW

File: src/iter.rs [hunk 1]
```diff
@@ -222,6 +225,19 @@ pub trait Reads: Send + Sync {
         )
     }
 
+    #[must_use]
+    fn norm(
+        self,
+        selector_expr: SelectorExpr,
+        label: Label,
+        range: (usize, usize),
```
Comment (Daniel-Liu-c0deb0t) [inline src/iter.rs]: Use a generic bounded by [RangeBounds](https://doc.rust-lang.org/std/ops/trait.RangeBounds.html) for Rust's range syntax (eg., `1..3`)! See the `length_in_bounds` operation for my usage.

Comment (noahcape) [reply src/iter.rs]: I want to be able to get both of the two number separately. Don't think there is convenient way to do that with RangeBounds. Let me know if I'm wrong.

File: src/iter/normalize_reads.rs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+pub struct NormalizeReads<R: Reads> {
+    reads: R,
+    selector_expr: SelectorExpr,
+    label: Label,
+    range: (usize, usize),
+}
+
+pub const NUC_MAP: [u8; 4] = [b'A', b'C', b'T', b'G'];
+
+pub fn log2_roundup(n: usize) -> usize {
```
Comment (Daniel-Liu-c0deb0t) [inline src/iter/normalize_reads.rs]: You can probably use [`leading_zeros`](https://stackoverflow.com/questions/72251467/computing-ceil-of-log2-in-rust)?

File: src/iter.rs (hunk 1 above)
Comment (Daniel-Liu-c0deb0t) [reply src/iter.rs]: Does [this](https://doc.rust-lang.org/std/ops/trait.RangeBounds.html#tymethod.start_bound) work?

Comment (noahcape) [reply src/iter.rs]: This is what I saw but requires unpacking from the Bound. If you think this is cleaner let me know.

Comment (Daniel-Liu-c0deb0t) [reply src/iter.rs]: Yeah its annoying to unpack, but I would prefer to have the nice `..` syntax for the API.

File: src/read.rs [hunk 3]
```diff
@@ -267,6 +268,47 @@ impl StrMappings {
...
+        let mut length_diff = long_len - normalized.len;
+
+        let extra_len = log4_roundup(long_len - short_len + 1);
+
+        let normed_len = long_len - normalized.len + extra_len;
+
+        self.mappings.iter_mut().for_each(|m| {
+            use Intersection::*;
+            match normalized.intersect(m) {
+                BAOverlap(_) | ABOverlap(_) | AInsideB | ABeforeB | Equal => m.len += normed_len,
```
Comment (noahcape) [inline src/read.rs:287]: I actually think this needs to be thought through differently. May need a different metric to decide padding.

Comment (Daniel-Liu-c0deb0t) [reply src/read.rs:287]: You may be able to directly use `set` to set an interval to a different string: https://github.com/Daniel-Liu-c0deb0t/ANTISEQUENCE/blob/main/src/read.rs#L613

Comment (Daniel-Liu-c0deb0t) [reply src/read.rs:287]: `set` will adjust all the intervals to account for the modified substring.

Comment (noahcape) [reply src/read.rs:287]: I think I did the least disruptive approach in the new pr for `pad`

**Pull request closed without merging**